File: translate_json.py
#!/usr/bin/env python3
"""
Auto-translate JSON translation files using Google Cloud Translate API
Usage: python translate_json.py
"""

import json
import os
from google.cloud import translate_v2 as translate

# Language mappings
LANGUAGES = {
    'de': 'German',
    'fr': 'French',
    'es': 'Spanish',
    'lt': 'Lithuanian',
    'el': 'Greek'
}

def translate_text(text, target_language, translate_client):
    """Translate text to target language"""
    if not text or text.strip() == "":
        return text
    
    try:
        result = translate_client.translate(
            text,
            target_language=target_language,
            source_language='en'
        )
        return result['translatedText']
    except Exception as e:
        print(f"Error translating '{text}' to {target_language}: {e}")
        return text

def translate_dict(data, target_language, translate_client, path=""):
    """Recursively translate all string values in a dictionary"""
    if isinstance(data, dict):
        result = {}
        for key, value in data.items():
            current_path = f"{path}.{key}" if path else key
            print(f"Translating: {current_path}")
            result[key] = translate_dict(value, target_language, translate_client, current_path)
        return result
    elif isinstance(data, str):
        # Skip if string contains only special characters or is a placeholder
        if data.strip() and not data.startswith('{') and not data.endswith('}'):
            translated = translate_text(data, target_language, translate_client)
            print(f"  '{data}' -> '{translated}'")
            return translated
        return data
    else:
        return data

def main():
    # Initialize Translate client
    print("Initializing Google Cloud Translate API...")
    translate_client = translate.Client()
    
    # Get the assets/translations directory
    script_dir = os.path.dirname(os.path.abspath(__file__))
    translations_dir = os.path.join(script_dir, 'assets', 'translations')
    
    # Read English JSON
    en_file = os.path.join(translations_dir, 'en.json')
    print(f"\nReading English translations from: {en_file}")
    
    with open(en_file, 'r', encoding='utf-8') as f:
        en_data = json.load(f)
    
    print(f"Found {len(en_data)} top-level keys to translate\n")
    
    # Translate to each language
    for lang_code, lang_name in LANGUAGES.items():
        print(f"\n{'='*60}")
        print(f"Translating to {lang_name} ({lang_code})...")
        print(f"{'='*60}\n")
        
        # Translate the entire structure
        translated_data = translate_dict(en_data, lang_code, translate_client)
        
        # Save to file
        output_file = os.path.join(translations_dir, f'{lang_code}.json')
        print(f"\nSaving to: {output_file}")
        
        with open(output_file, 'w', encoding='utf-8') as f:
            json.dump(translated_data, f, ensure_ascii=False, indent=2)
        
        print(f"✅ {lang_name} translation complete!")
    
    print(f"\n{'='*60}")
    print("🎉 ALL TRANSLATIONS COMPLETE!")
    print(f"{'='*60}")
    print("\nTranslated files:")
    for lang_code, lang_name in LANGUAGES.items():
        print(f"  ✅ {lang_name}: assets/translations/{lang_code}.json")

if __name__ == "__main__":
    main()

File: translate_json.js
/**
 * Auto-translate JSON translation files using Google Cloud Translate API
 * Usage: node translate_json.js
 */

const fs = require('fs').promises;
const path = require('path');
const { Translate } = require('@google-cloud/translate').v2;

// Language mappings
const LANGUAGES = {
  'de': 'German',
  'fr': 'French',
  'es': 'Spanish',
  'lt': 'Lithuanian',
  'el': 'Greek'
};

// Initialize Translate client
const translate = new Translate();

/**
 * Translate text to target language
 */
async function translateText(text, targetLanguage) {
  if (!text || text.trim() === "") {
    return text;
  }
  
  try {
    const [translation] = await translate.translate(text, {
      from: 'en',
      to: targetLanguage
    });
    return translation;
  } catch (error) {
    console.error(`Error translating '${text}' to ${targetLanguage}:`, error.message);
    return text;
  }
}

/**
 * Recursively translate all string values in an object
 */
async function translateDict(data, targetLanguage, path = "") {
  if (typeof data === 'object' && data !== null && !Array.isArray(data)) {
    const result = {};
    for (const [key, value] of Object.entries(data)) {
      const currentPath = path ? `${path}.${key}` : key;
      console.log(`Translating: ${currentPath}`);
      result[key] = await translateDict(value, targetLanguage, currentPath);
    }
    return result;
  } else if (typeof data === 'string') {
    // Skip if string contains only special characters or is a placeholder
    if (data.trim() && !data.startsWith('{') && !data.endsWith('}')) {
      const translated = await translateText(data, targetLanguage);
      console.log(`  '${data}' -> '${translated}'`);
      return translated;
    }
    return data;
  } else {
    return data;
  }
}

async function main() {
  console.log("Initializing Google Cloud Translate API...\n");
  
  // Get the assets/translations directory
  const translationsDir = path.join(__dirname, 'assets', 'translations');
  
  // Read English JSON
  const enFile = path.join(translationsDir, 'en.json');
  console.log(`Reading English translations from: ${enFile}\n`);
  
  const enData = JSON.parse(await fs.readFile(enFile, 'utf-8'));
  console.log(`Found ${Object.keys(enData).length} top-level keys to translate\n`);
  
  // Translate to each language
  for (const [langCode, langName] of Object.entries(LANGUAGES)) {
    console.log('='.repeat(60));
    console.log(`Translating to ${langName} (${langCode})...`);
    console.log('='.repeat(60) + '\n');
    
    // Translate the entire structure
    const translatedData = await translateDict(enData, langCode);
    
    // Save to file
    const outputFile = path.join(translationsDir, `${langCode}.json`);
    console.log(`\nSaving to: ${outputFile}`);
    
    await fs.writeFile(
      outputFile,
      JSON.stringify(translatedData, null, 2),
      'utf-8'
    );
    
    console.log(`✅ ${langName} translation complete!\n`);
  }
  
  console.log('='.repeat(60));
  console.log('🎉 ALL TRANSLATIONS COMPLETE!');
  console.log('='.repeat(60));
  console.log('\nTranslated files:');
  for (const [langCode, langName] of Object.entries(LANGUAGES)) {
    console.log(`  ✅ ${langName}: assets/translations/${langCode}.json`);
  }
}

main().catch(console.error);
